frontend/src/metabase/query_builder/components/view/ViewHeader/components/SavedQuestionLeftSide/util.ts

```diff
@@ -0,0 +1,21 @@
+import { useGetCardQueryMetadataQuery } from "metabase/api";
+import * as Lib from "metabase-lib";
+import type Question from "metabase-lib/v1/Question";
+import type { Table } from "metabase-types/api";
+
+export function useHiddenSourceTables(question: Question): Table[] {
+  const query = question.query();
+  const { isEditable, isNative } = Lib.queryDisplayInfo(query);
+
+  const canHaveHiddenSourceTables = !isEditable && !isNative;
+
+  const { data } = useGetCardQueryMetadataQuery(question.id(), {
+    skip: !canHaveHiddenSourceTables,
+  });
+
+  if (!canHaveHiddenSourceTables) {
+    return [];
+  }
+
+  return data?.tables.filter(table => table.visibility_type === "hidden") ?? [];
```

Comment:
I believe that this approach - where you check each table in card query metadata - might not work because extra tables can be returned. For instance, an implicitly joinable table can be hidden but it is not used in the query and thus it would be incorrect to use it here.

E.g. a possible bug:
- data source is a inaccessible card
- there are accessible (and visible) joined tables
- there are hidden implicitly joined tables

The query is readonly because the source card is unavailable but this logic would show the name of the first implicitly joinable table.

Reply:
Unsure if that's a correct approach but what you could do:

First of all, you need to know that `getMetadata` selector removes hidden tables by default so you won't get them with MBQL lib. You could re-construct the `query` object with unfiltered metadata:
```
const datasetQuery = question.datasetQuery();
const metadata = useSelector(getMetadataUnfiltered);
const metadataProvider = Lib.metadataProvider(datasetQuery.database, metadata);
const query = Lib.fromLegacyQuery(datasetQuery.database, metadataProvider, datasetQuery);

const sourceTableId = Lib.sourceTableOrCardId(query);
const sourceTable = Lib.tableOrCardMetadata(sourceTableId);
const sourceTableInfo = sourceTable ? Lib.displayInfo(query, -1, sourceTable) : undefined;
const joinTablesInfo = Lib.stageIndexes(query).flatMap(stageIndex => 
  Lib.joins(query, stageIndex)
       .map(join => Lib.joinedThing(query, join))
       .filter(joinTable => joinTable != null)
       .map(joinTable => Lib.displayInfo(query, stageIndex, joinTable));

const tablesInfo = [
   ...(sourceTableInfo ? [sourceTableInfo] : []),
   ...joinTablesInfo
];
return tablesInfo.first(tableInfo => tableInfo.isHidden);
```

^ This assumes that:
- There is a new property in `TableDisplayInfo` in MBQL lib. You could add one in clojure based on `visibility-type` of the table.
- There is no need to add an API call

Reply:
Also - let's not merge non-trivial changes to master without a testing plan.

Reply:
I've implemented this suggestion and it works!

Not sure I would've ever been able to come up with that code snippet myself (let alone believe that it would work)...

I've added the extra cases to the [testing plan](https://www.notion.so/metabase/Testing-plan-Disallow-editing-queries-with-hidden-tables-12369354c90180f9be40c4bea7fba5ba) and wrote the tests too.
Am I missing any test cases?